core/src/main/java/io/undertow/server/handlers/proxy/mod_cluster/ModClusterContainer.java

```diff
@@ -392,12 +393,47 @@ Context findNewNode(final VirtualHost.HostEntry entry) {
     /**
      * Try to find a failover node within the same load balancing group.
      *
-     * @param domain   the load balancing domain, if known
-     * @param jvmRoute the original jvmRoute
+     * @param entry              the resolved virtual host entry
+     * @param domain             the load balancing domain, if known
+     * @param session            the actual value of JSESSIONID/jsessionid cookie/parameter
+     * @param jvmRoute           the original jvmRoute
+     * @param forceStickySession whether sticky sessions are forced
      * @return the context, {@code null} if not found
-     * @oaram entry      the resolved virtual host entry
      */
-    Context findFailoverNode(final VirtualHost.HostEntry entry, final String domain, final String jvmRoute, final boolean forceStickySession) {
+    Context findFailoverNode(final VirtualHost.HostEntry entry, final String domain, final String session, final String jvmRoute, final boolean forceStickySession) {
+
+        // If configured, deterministically choose the failover target by calculating hash of the session ID modulo number of electable nodes
+        if (modCluster.isDeterministicFailover()) {
+            List<String> candidates = new ArrayList<>(entry.getNodes().size());
+            for (String route : entry.getNodes()) {
+                Node node = nodes.get(route);
+                if (node != null && !node.isInErrorState() && !node.isHotStandby()) {
+                    candidates.add(route);
+                }
+            }
+
+            // If there are no available regular nodes, all hot standby nodes become candidates
+            if (candidates.isEmpty()) {
+                for (String route : entry.getNodes()) {
+                    Node node = nodes.get(route);
+                    if (node != null && !node.isInErrorState() && node.isHotStandby()) {
+                        candidates.add(route);
+                    }
+                }
+            }
+
+            if (candidates.isEmpty()) {
+                return null;
+            }
+
+            String sessionId = session.substring(0, session.indexOf('.'));
+            int index = Math.abs(sessionId.hashCode()) % candidates.size();
```

Comment:
Findbugs is complaining about this:

RV: Bad attempt to compute absolute value of signed 32-bit hashcode (RV_ABSOLUTE_VALUE_OF_HASHCODE)

This code generates a hashcode and then computes the absolute value of that hashcode. If the hashcode is Integer.MIN_VALUE, then the result will be negative as well (since Math.abs(Integer.MIN_VALUE) == Integer.MIN_VALUE).

One out of 2^32 strings have a hashCode of Integer.MIN_VALUE, including "polygenelubricants" "GydZG_" and ""DESIGNING WORKHOUSES".